samples/siwx91x_ota/src/main.c

```diff
@@ -0,0 +1,1029 @@
+		if (status != SL_STATUS_OK) {
+			printf("Failed to load firmware header (0x%x)\n", status);
+			return status;
+		}
+
+		/* Load first chunk (after header) */
+		status = sl_si91x_fwup_load((otaf_image + ota_header_size),
+					    (length - ota_header_size));
+		if (status != SL_STATUS_OK) {
+			printf("Failed to load first firmware chunk (0x%x)\n", status);
+			return status;
+		}
+
+		/* Mark that we've processed the first packet */
+		first_packet = false;
+		return SL_STATUS_OK;
+	}
+
+	/* Handle subsequent packets */
+	status = sl_si91x_fwup_load(otaf_image, length);
+
+	/* Check if firmware update is completed */
+	if (status == SL_STATUS_SI91X_FW_UPDATE_DONE) {
+		printf("Firmware update completed\n");
+		zsock_close(app_ctx_st->sock);
+		osDelay(3000);
+		printf("SoC Soft Reset initiated!\n");
+		sys_arch_reboot();
+
+		/* Reset first_packet for next update session */
+		first_packet = true;
+		return SL_STATUS_OK;
+	} else if (status != SL_STATUS_OK) {
+		printf("Fail to load the firmware chunk: 0x%x\n", status);
+		return status;
+	}
+	return SL_STATUS_OK;
+}
+
+/*
+ * @brief Process downloaded firmware chunk
+ * @param app_ctx_st Pointer to application context
+ * @return int 0 if more chunks needed, 1 if complete, negative on error
+ */
+static int process_firmware_chunk(struct app_ctx *app_ctx_st)
+{
+	int status;
+	int req_start;
+
+	printf("Processing firmware data...\n");
+
+	/* Wait for HTTP response to be fully received */
+	if (k_sem_take(&app_ctx_st->ota_sem, K_MSEC(FW_CHUNK_DOWNLOAD_TIMEOUT_MS)) != 0) {
+		printf("Firmware download failed or Semaphore timed out\n");
+		return -errno;
+	}
+
+	/* Check if there was an error in the HTTP response */
+	if (app_ctx_st->http_response_error) {
+		printf("Error occurred during HTTP download, skipping processing\n");
+		return -EPROTO;
+	}
+	req_start = app_ctx_st->ota_range_start_byte;
+
+	/* Process all queued data chunks */
+	while (k_msgq_get(&app_ctx_st->http_data_q, &app_ctx_st->msg_c, K_NO_WAIT) == 0) {
+		app_ctx_st->ota_range_start_byte =
+			app_ctx_st->ota_range_start_byte + app_ctx_st->msg_c.length;
+		status = ota_load_firmware(app_ctx_st);
+		if (status != SL_STATUS_OK) {
+			printf("Fw load failed for requested range=%d-%d\n", req_start,
+			       app_ctx_st->ota_range_end_byte);
+			app_ctx_st->ota_range_start_byte = req_start;
+			return -EIO;
+		}
+	}
+	printf("Fw load success for requested range=%d-%d\n", req_start,
+	       app_ctx_st->ota_range_end_byte);
+	return 0;
+}
+
+/*
+ * @brief Prints the current firmware version
+ *
+ * Retrieves and displays the current firmware version information
+ * from the device.
+ */
+void ota_print_firmware_version(void)
+{
+	int ret = 0;
+	sl_wifi_firmware_version_t version = {};
+
+	/* Get initial firmware version */
+	ret = sl_wifi_get_firmware_version(&version);
+	if (ret == SL_STATUS_OK) {
+		print_firmware_version(&version);
+	} else {
+		printf("Failed to get firmware version: 0x%x\n", ret);
+	}
+}
```

Comment:
Does it make sense to provide this information while we are not able to provide the version of the M4 firmware?